dna/zomes/courses/code/src/course.rs

```diff
@@ -157,82 +261,173 @@ pub fn anchor_entry_def() -> ValidatingEntryType {
+    hdk::link_entries(
+        &course_anchor_address,
+        &course_data_address,
+        "course_anchor->course_data",
+        "",
+    )?;
 
-    Ok(new_course_address)
+    Ok(course_anchor_address)
 }
 
 pub fn update(
     title: String,
     modules_addresses: Vec<Address>,
     course_address: &Address,
 ) -> ZomeApiResult<Address> {
-    let course: Course = hdk::utils::get_as_type(course_address.clone())?;
+    //TODO:  check if addr we receive is actually CourseAnchor, not CourseData
 
-    let new_version_course = Course::from(
+    // get latest CourseData entry by looking at links of CourseAnchor entry
+    let latest_course_data_addr =
+        get_latest_link_addr(course_address, "course_anchor->course_data")?;
+    let course_data: CourseData = hdk::utils::get_as_type(latest_course_data_addr)?;
+
+    // create a new version of CourseData entry
+    let new_course_data_entry = CourseData::from(
         title,
-        course.teacher_address,
-        course.timestamp,
+        course_data.teacher_address,
+        course_data.course_anchor,
+        // TODO: we shouldn't be using old timestamp here. Maybe we should receive it as a parameter for update?
+        course_data.timestamp,
         modules_addresses,
-    );
-    let new_version_course_entry = new_version_course.entry();
+    )
+    .entry();
+    let new_course_data_address = hdk::commit_entry(&new_course_data_entry)?;
 
-    hdk::update_entry(new_version_course_entry, course_address)
+    hdk::link_entries(
+        &course_address,
+        &new_course_data_address,
+        "course_anchor->course_data",
+        "",
+    )?;
+
+    // since CourseAnchor entry stays the same, we're not returning any new addresses here
+    // and since we don't have ownership of the course_address in this method, we're cloning it
+    // to comply with the return type requirements
+    Ok(course_address.clone())
 }
 
-pub fn delete(address: Address) -> ZomeApiResult<Address> {
-    hdk::remove_link(&anchor_address()?, &address, "course_list", "")?;
+pub fn delete(course_address: Address) -> ZomeApiResult<Address> {
+    //TODO:  check if addr we receive is actually CourseAnchor, not CourseData
+
+    // remove link from all_courses anchor to this course's CourseAnchor entry
+    hdk::remove_link(
+        &anchor_all_courses_address()?,
+        &course_address,
+        "course_list",
+        "",
+    )?;
+
+    // find addresses of all CourseData entries for the course
+    let course_data_addresses = hdk::get_links(
+        &course_address,
+        LinkMatch::Exactly("course_anchor->course_data"),
+        LinkMatch::Any,
+    )?
+    .addresses();
 
-    let students = get_students(address.clone())?;
-    let course: Course = hdk::utils::get_as_type(address.clone())?;
+    // remove all CourseData for the course
+    for course_data_addr in course_data_addresses {
+        hdk::remove_entry(&course_data_addr)?;
```

Comment:
Necessary? It will only add new entries to the DHT (won't actually clean up any memory, only add new things to hold and so on) and these course_data entries will be basically inaccessible once we delete the course_anchor